### PR TITLE
CFY-6201 integration of aria models into cloudify

### PR DESCRIPTION
**Pull request by mxmrlv**

### CODE REVIEW

File: rest-service/dev-requirements.txt [hunk 1]
```diff
@@ -1,5 +1,7 @@
 https://github.com/cloudify-cosmo/cloudify-dsl-parser/archive/master.zip
 https://github.com/cloudify-cosmo/cloudify-rest-client/archive/master.zip
+https://github.com/apache/incubator-ariatosca/archive/ARIA-39-Genericize-storage-models.zip
```
Comment (mcouthon) [inline rest-service/dev-requirements.txt]: Remove

File: rest-service/manager_rest/resource_manager.py [hunk 2]
```diff
@@ -602,7 +607,7 @@ def prepare_deployment_nodes_for_storage(self,
                 deploy_number_of_instances=scalable['default_instances'],
                 min_number_of_instances=scalable['min_instances'],
                 max_number_of_instances=scalable['max_instances'],
-                host_id=raw_node['host_id'] if 'host_id' in raw_node else None,
+                host_fk=host_list[0].storage_id if host_list else None,
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: replace host_fk = a.storage_id with host=a

File: rest-service/manager_rest/resource_manager.py [hunk 3]
```diff
@@ -619,10 +624,14 @@ def _prepare_deployment_node_instances_for_storage(self,
             instance_id = node_instance['id']
             scaling_groups = node_instance.get('scaling_groups', [])
             relationships = node_instance.get('relationships', [])
-            host_id = node_instance.get('host_id')
+            host_list = self.sm.list(models.NodeInstance,
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: get instead of list

File: rest-service/manager_rest/resource_manager.py [hunk 4]
```diff
@@ -593,7 +592,13 @@ def prepare_deployment_nodes_for_storage(self,
         nodes = []
         for raw_node in raw_nodes:
             scalable = raw_node['capabilities']['scalable']['properties']
+            host_list = self.sm.list(models.Node,
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: There is a get_node method in the resource manager. Use it 🗡

File: rest-service/manager_rest/resource_manager.py [hunk 5]
```diff
@@ -638,12 +647,16 @@ def _create_deployment_nodes(self,
...
         deployment = self.sm.get(models.Deployment, deployment_id)
 
+        nodes = self.prepare_deployment_nodes_for_storage(
+            plan, deployment, node_ids)
+
         for node in nodes:
-            node.deployment = deployment
             self.sm.put(node)
+            if node.host is None:
+                node.host_fk = node.storage_id
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: ditto

File: rest-service/manager_rest/resource_manager.py [hunk 6]
```diff
@@ -654,6 +667,9 @@ def _create_deployment_node_instances(self,
 
         for node_instance in node_instances:
             self.sm.put(node_instance)
+            if node_instance.host is None:
+                node_instance.host_fk = node_instance.storage_id
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: ditto

File: rest-service/manager_rest/resource_manager.py [hunk 7]
```diff
@@ -654,6 +667,9 @@ def _create_deployment_node_instances(self,
 
         for node_instance in node_instances:
             self.sm.put(node_instance)
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: Move below the if

File: rest-service/manager_rest/resource_manager.py [hunk 8]
```diff
@@ -638,12 +647,16 @@ def _create_deployment_nodes(self,
...
                                  node_ids=None):
-        nodes = self.prepare_deployment_nodes_for_storage(plan, node_ids)
         deployment = self.sm.get(models.Deployment, deployment_id)
 
+        nodes = self.prepare_deployment_nodes_for_storage(
+            plan, deployment, node_ids)
+
         for node in nodes:
-            node.deployment = deployment
             self.sm.put(node)
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: Move below the if

File: rest-service/manager_rest/resource_manager.py [hunk 9]
```diff
@@ -912,14 +928,20 @@ def add_node_instance_from_dict(self, instance_dict):
         node_id = instance_dict.pop('node_id')
 
         # Link the node instance object to to the node, and add it to the DB
-        new_node_instance = models.NodeInstance(**instance_dict)
+        host_id = instance_dict.pop('host_id')
+        hosts = self.sm.list(models.NodeInstance, filters={'id': host_id})
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: Create a get_node_instance method

File: rest-service/manager_rest/resource_manager.py [hunk 10]
```diff
@@ -912,14 +928,20 @@ def add_node_instance_from_dict(self, instance_dict):
...
+        new_node_instance = models.NodeInstance(
+            host=hosts[0] if hosts else None, **instance_dict)
```
Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: Add to dict instead

Comment (mcouthon) [inline rest-service/manager_rest/resource_manager.py]: And add itself if None

File: rest-service/manager_rest/security/user_handler.py [hunk 11]
```diff
@@ -26,7 +26,7 @@ def user_loader(request):
     Having this function makes sure that this will work:
     > from flask_security import current_user
     > current_user
-    <manager_rest.storage.models.User object at 0x50d9d10>
+    <manager_rest.storage.resource_models.py.User object at 0x50d9d10>
```
Comment (mcouthon) [inline rest-service/manager_rest/security/user_handler.py]: .py

File: rest-service/manager_rest/storage/management_models.py [hunk 12]
```diff
@@ -80,13 +74,14 @@ class Role(SQLModelBase, RoleMixin):
...
 
-    def _get_identifier(self):
-        return 'name', self.name
-
 
 class User(SQLModelBase, UserMixin):
     __tablename__ = 'users'
 
+    @classmethod
+    def name_column_name(cls):
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/management_models.py]: explain WTF that is

File: rest-service/manager_rest/storage/resource_models.py [hunk 13]
```diff
@@ -20,96 +20,81 @@
...
     created_at = db.Column(UTCDateTime, nullable=False, index=True)
-    main_file_name = db.Column(db.Text, nullable=False)
-    plan = db.Column(db.PickleType, nullable=False)
-    updated_at = db.Column(UTCDateTime)
-    description = db.Column(db.Text)
+    updated_at = db.Column(UTCDateTime, index=True)
+
+    # old style id support
+    @declared_attr
+    def blueprint_id(cls):
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: Should be in deployment, maybe? Fo sho!

File: rest-service/manager_rest/storage/resource_models.py [hunk 14]
```diff
@@ -20,96 +20,81 @@
...
-    excluded_wheels = db.Column(db.PickleType)
-    package_name = db.Column(db.Text, nullable=False, index=True)
-    package_source = db.Column(db.Text)
-    package_version = db.Column(db.Text)
-    supported_platform = db.Column(db.PickleType)
-    supported_py_versions = db.Column(db.PickleType)
+class Plugin(base.PluginBase, TopLevelResource):
     uploaded_at = db.Column(UTCDateTime, nullable=False, index=True)
-    wheels = db.Column(db.PickleType, nullable=False)
+    excluded_wheels = db.Column(aria_types.List)
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: Should be in Aria?

File: rest-service/manager_rest/storage/resource_models.py [hunk 15]
```diff
@@ -119,9 +104,12 @@ def parent(self):
     def parent(cls):
         return Blueprint
 
-    blueprint_id = association_proxy('blueprint', 'id')
     tenant_id = association_proxy('blueprint', 'tenant_id')
 
+    @declared_attr
+    def blueprint_id(cls):
+        return association_proxy('blueprint', cls.name_column_name())
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: Change to 'id' string

File: rest-service/manager_rest/storage/resource_models.py [hunk 16]
```diff
@@ -181,15 +165,24 @@ def parent(self):
...
             id_name,
-            id_value,
+            getattr(self, id_name),
             self.status
         )
 
+    # old style id support
+    @declared_attr
+    def blueprint_id(cls):
+        return association_proxy('deployment', 'blueprint_id')
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: No need for a method - can be a one liner

File: rest-service/manager_rest/storage/resource_models.py [hunk 17]
```diff
@@ -181,15 +165,24 @@ def parent(self):
...
+
+    @declared_attr
+    def deployment_id(cls):
+        return association_proxy('deployment', cls.name_column_name())
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: ditto

File: rest-service/manager_rest/storage/resource_models.py [hunk 18]
```diff
@@ -294,36 +273,37 @@ def parent(self):
...
         return Deployment
 
-    deployment_id = association_proxy('deployment', 'id')
-    execution_id = association_proxy('execution', 'id')
     tenant_id = association_proxy('deployment', 'tenant_id')
 
+    # old style id support
+    @declared_attr
+    def execution_id(cls):
+        return association_proxy('execution', cls.name_column_name())
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: ditto

File: rest-service/manager_rest/storage/resource_models.py [hunk 19]
```diff
@@ -294,36 +273,37 @@ def parent(self):
...
+
+    @declared_attr
+    def deployment_id(cls):
+        return association_proxy('deployment', cls.name_column_name())
+
+    @declared_attr
+    def blueprint_id(cls):
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: Remove (I think it's unused)

File: rest-service/manager_rest/storage/resource_models.py [hunk 20]
```diff
@@ -419,13 +372,27 @@ def parent(self):
...
+    @declared_attr
+    def host_id(cls):
+        return association_proxy('host', cls.name_column_name())
+
+    @declared_attr
+    def deployment_id(cls):
+        return association_proxy('deployment', cls.name_column_name())
+
+    @declared_attr
+    def blueprint_id(cls):
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: נו מה

File: rest-service/manager_rest/storage/resource_models.py [hunk 21]
```diff
@@ -434,35 +401,59 @@ class NodeInstance(DerivedResource, DerivedMixin):
...
     @declared_attr
-    def node(cls):
-        return one_to_many_relationship(cls, Node, cls.node_fk)
+    def host_id(cls):
+        return association_proxy('host', cls.name_column_name())
+
+    relationships = db.Column(aria_types.List)
+
+
+class Task(base.TaskBase, DerivedResource, DerivedMixin):
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: Add a comment to explain what it is

File: rest-service/manager_rest/storage/resource_models.py [hunk 22]
```diff
@@ -434,35 +401,59 @@ class NodeInstance(DerivedResource, DerivedMixin):
...
     def parent(cls):
-        return Node
+        return Execution
 
-    node_id = association_proxy('node', 'id')
-    deployment_id = association_proxy('node', 'deployment_id')
-    tenant_id = association_proxy('node', 'tenant_id')
+    # old style id support
+    @declared_attr
+    def execution_id(cls):
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models.py]: remove

File: rest-service/manager_rest/storage/resource_models_base.py [hunk 23]
```diff
@@ -32,8 +32,20 @@ class SQLResourceBase(SQLModelBase):
...
+    @classmethod
+    def name_column_name(cls):
+        return 'id'
+
+    storage_id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    id = db.Column(db.Text, index=True)
+
     # Differentiates between resources (blueprints, nodes, etc.) and other
-    # table models (users, tenants, etc.)
+    # table resource_models.py (users, tenants, etc.)
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/resource_models_base.py]: .py

File: rest-service/manager_rest/storage/storage_manager.py [hunk 24]
```diff
@@ -380,7 +380,10 @@ def _validate_unique_resource_id_per_tenant(self, instance):
         if not instance.is_resource or not instance.is_id_unique:
             return
 
-        filters = {'id': instance.id, 'tenant_id': self.current_tenant.id}
+        filters = {
```
Comment (mcouthon) [inline rest-service/manager_rest/storage/storage_manager.py]: Add comment
 מבסוט 🥇

File: rest-service/manager_rest/test/endpoints/test_deployment_updates.py [hunk 25]
```diff
@@ -276,33 +277,34 @@ def test_one_active_update_per_deployment_force_flag(self):
...
-        self.assertEquals(response.json['error_code'], 'conflict_error')
-        self.assertIn('there are deployment updates still active',
-                      response.json['message'])
-        self.assertIn('the "force" flag was used',
-                      response.json['message'])
-        # verifying the first update wasn't set with a failed state by the
-        # force flag call
-        first_update = self.client.deployment_updates.get(first_update_id)
-        self.assertEquals(STATES.EXECUTING_WORKFLOW, first_update.state)
+        with validate_execution_transitions():
```
Comment (mcouthon) [inline rest-service/manager_rest/test/endpoints/test_deployment_updates.py]: Change name to dont validate or validate all or skip validation or maybe you shouldn't really check transition validation please sir goodday

File: rest-service/manager_rest/test/mocks.py [hunk 26]
```diff
@@ -71,7 +72,8 @@ def do_request(self,
         )
 
     def _do_request(self, requests_method, request_url, body, params, headers,
-                    expected_status_code, stream, verify, timeout=None):
+                    expected_status_code, stream, verify, timeout=None,
+                    **kwargs):
```
Comment (mcouthon) [inline rest-service/manager_rest/test/mocks.py]: Check if can be removed

File: rest-service/manager_rest/test/storage/base_test.py [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+@attr(client_min_version=1, client_max_version=LATEST_API_VERSION)
+class ModelsIntegrationTestsBase(unittest.TestCase):
+    def setUp(self):
+        rest_service_log, tmp_conf_file = \
+            self._create_temp_files_and_folders()
+        self.server = self._set_config_path_and_get_server_module(
+            rest_service_log, tmp_conf_file)
+        self.server_configuration = self._create_config_and_reset_app(
+            rest_service_log)
+        self._set_flask_app_context()
```
Comment (mcouthon) [inline rest-service/manager_rest/test/storage/base_test.py]: Create a test base class in tests/base_test.py that doesn't use clients and stuff.
BallerTestCase

File: rest-service/manager_rest/test/storage/base_test.py [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+        """Create config, and reset Flask app
+        :type server: module
+        """
+        server_configuration = self._create_configuration(rest_service_log)
+        self.server.SQL_DIALECT = 'sqlite'
+        self.server.reset_app(server_configuration)
+        return server_configuration
+
+    @staticmethod
+    def _create_configuration(rest_service_log):
```
Comment (mcouthon) [inline rest-service/manager_rest/test/storage/base_test.py]: Make a base class for both base tests classes.

File: rest-service/manager_rest/test/storage/base_test.py [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+        finally:
+            del(os.environ['MANAGER_REST_CONFIG_PATH'])
+        return server
+
+    def _set_flask_app_context(self):
+        flask_app_context = self.server.app.test_request_context()
+        flask_app_context.push()
+        self.addCleanup(flask_app_context.pop)
+
+    def _init_default_tenant(self):
```
Comment (mcouthon) [inline rest-service/manager_rest/test/storage/base_test.py]: Unacceptable

File: rest-service/manager_rest/test/utils.py [hunk 30]
```diff
@@ -59,3 +63,44 @@ def tar_file(file_to_tar, destination_dir, tar_name=''):
...
+
+    def _exit(self):
+        models.Execution.VALID_TRANSITIONS.clear()
+        models.Execution.VALID_TRANSITIONS.update(
+            self._original_valid_transition)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._exit()
+
+    def __call__(self, func=None, new_transition=None, **kwargs):
```
Comment (mcouthon) [inline rest-service/manager_rest/test/utils.py]: יפה

**Comment by geokala** [comment]
Closed: As this PR is more than six months old without having been merged. Please reopen (and rebase if necessary) if it is still relevant.

**Pull request closed without merging**